# DEALING WITH MISSING VALUES

<span style="color: #696969;">In this tutorial, you will learn three approaches to dealing with missing values. Then you'll compare the effectiveness of these approaches on a real-world dataset.</span>

# Introduction

<span style="color: #696969;">
    
There are many ways data can end up with missing values. For example,
+ A 2 bedroom house won't include a value for the size of a third bedroom.
+ A survey respondent may choose not to share his income.
    
Most machine learning libraries (including scikit-learn) give an error if you try to build a model using data with missing values. So you'll need to choose one of the strategies below.

</span>

# Five Approaches

<b>1. A Simple Option: Drop Columns with Missing Values</b>

<span style="color: #696969;">The simplest option is to drop columns with missing values.</span>

![Drop values](https://github.com/bharathkreddy/ML-Bootcamp/blob/master/img/image1.png)

<span style="color: #696969;">
Unless most values in the dropped columns are missing, the model loses access to a lot of (potentially useful!) information with this approach. As an extreme example, consider a dataset with 10,000 rows, where one important column is missing a single entry. This approach would drop the column entirely!
</span>

<b>2. A Better Option: Imputation</b>

<span style="color: #696969;">
Imputation fills in the missing values with some number. For instance, we can fill in the mean value along each column.
</span>

![impute](https://github.com/bharathkreddy/ML-Bootcamp/blob/master/img/image2.png)

<span style="color: #696969;">
The imputed value won't be exactly right in most cases, but it usually leads to more accurate models than you would get from dropping the column entirely.
</span>

<b>3. An Extension To Imputation</b>


<span style="color: #696969;">
Imputation is the standard approach, and it usually works well. However, imputed values may be systematically above or below their actual values (which weren't collected in the dataset). Or rows with missing values may be unique in some other way. In that case, your model would make better predictions by considering which values were originally missing.
</span>

![impute](https://github.com/bharathkreddy/ML-Bootcamp/blob/master/img/image3.png)


<span style="color: #696969;">
In this approach, we impute the missing values, as before. And, additionally, for each column with missing entries in the original dataset, we add a new column that shows the location of the imputed entries.
  
In some cases, this will meaningfully improve results. In other cases, it doesn't help at all.
    
</span>


<b>4. Stratified mean</b>

<span style="color: #696969;">Break the dataset into parts based on strata and impute mean of each stata, use this mean to fill mising values for each of the strata. Allways use this if data is stratified.</span>

<b>5. KNN imputer : this uses nearest K neighbouring values to impute the missing value</b>

<span style="color: #696969;">Use this option where data is very varied and no inherent structure to data.</span>

<span style="color: #696969;">

## Example

In the example, we will work with the [Melbourne Housing dataset](https://github.com/bharathkreddy/ML-Bootcamp/blob/master/data/melb_data.csv). Our model will use information such as the number of rooms and land size to predict home price.

First the data loading step. Split the data into training and validation data in `X_train`, `X_valid`, `y_train`, and `y_valid`. Where in - X are the `Features` and y is the `predicted variable`.
</span>

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Load the data
data = pd.read_csv('./data/melb_data.csv')

print(data)

# Select target
y = data.Price

# To keep things simple, we'll use only numerical predictors, we can do this by filtering out all the columns with dtypes = 'object'
melb_predictors = data.drop(['Price'], axis=1)
X = melb_predictors.select_dtypes(exclude=['object'])

# Divide data into training and validation subsets, note we use random_state for reproducability - google why its used
X_train, X_valid, y_train, y_valid = train_test_split(X, y, train_size=0.8, test_size=0.2,
                                                      random_state=0)

              Suburb           Address  Rooms Type      Price Method  \
0         Abbotsford      85 Turner St      2    h  1480000.0      S   
1         Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2         Abbotsford      5 Charles St      3    h  1465000.0     SP   
3         Abbotsford  40 Federation La      3    h   850000.0     PI   
4         Abbotsford       55a Park St      4    h  1600000.0     VB   
...              ...               ...    ...  ...        ...    ...   
13575  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
13576   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
13577   Williamstown       83 Power St      3    h  1170000.0      S   
13578   Williamstown      96 Verdon St      4    h  2500000.0     PI   
13579     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  \
0        Biggin   3/12/2016       2.5    3067.0  ...  

In [2]:
X_train

Rooms  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
12167      1       5.0    3182.0       1.0       1.0  1.0       0.0   
6524       2       8.0    3016.0       2.0       2.0  1.0     193.0   
8413       3      12.6    3020.0       3.0       1.0  1.0     555.0   
2919       3      13.0    3046.0       3.0       1.0  1.0     265.0   
6043       3      13.3    3020.0       3.0       1.0  2.0     673.0   
...      ...       ...       ...       ...       ...  ...       ...   
13123      3       5.2    3056.0       3.0       1.0  2.0     212.0   
3264       3      10.5    3081.0       3.0       1.0  1.0     748.0   
9845       4       6.7    3058.0       4.0       2.0  2.0     441.0   
10799      3      12.0    3073.0       3.0       1.0  1.0     606.0   
2732       4       6.4    3011.0       4.0       2.0  1.0     319.0   

       BuildingArea  YearBuilt  Lattitude  Longtitude  Propertycount  
12167           NaN     1940.0  -37.85984   144.98670        13240.0  
6524            NaN        NaN  -37.85800   144.90050         6380.0  
8413            NaN        NaN  -37.79880   144.82200         3755.0  
2919            NaN     1995.0  -37.70830   144.91580         8870.0  
6043          673.0     1970.0  -37.76230   144.82720         4217.0  
...             ...        ...        ...         ...            ...  
13123           NaN        NaN  -37.77695   144.95785        11918.0  
3264          101.0     1950.0  -37.74160   145.04810         2947.0  
9845          255.0     2002.0  -37.73572   144.97256        11204.0  
10799           NaN        NaN  -37.72057   145.02615        21650.0  
2732          130.0     1915.0  -37.79430   144.88750         7570.0  

[10864 rows x 12 columns]

## Define Function to Measure Quality of Each Approach



<span style="color: #696969;">
We define a function score_dataset() to compare different approaches to dealing with missing values. This function reports the mean absolute error (MAE) from a Linear Regression model.
</span>

In [3]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

# Function for comparing different approaches
def score_dataset(X_train, X_valid, y_train, y_valid):
    model = LinearRegression()
    model.fit(X_train, y_train)
    preds = model.predict(X_valid)
    return mean_absolute_error(y_valid, preds)

### Score from Approach 1 (Drop Columns with Missing Values)

Since we are working with both training and validation sets, we are careful to drop the same columns in both DataFrames.

### Get names of columns with missing values
### STEP 1: use .isnull() to find missing vlaues

In [4]:
X_train.isnull()

Rooms  Distance  Postcode  Bedroom2  Bathroom    Car  Landsize  \
12167  False     False     False     False     False  False     False   
6524   False     False     False     False     False  False     False   
8413   False     False     False     False     False  False     False   
2919   False     False     False     False     False  False     False   
6043   False     False     False     False     False  False     False   
...      ...       ...       ...       ...       ...    ...       ...   
13123  False     False     False     False     False  False     False   
3264   False     False     False     False     False  False     False   
9845   False     False     False     False     False  False     False   
10799  False     False     False     False     False  False     False   
2732   False     False     False     False     False  False     False   

       BuildingArea  YearBuilt  Lattitude  Longtitude  Propertycount  
12167          True      False      False       False          False  
6524           True       True      False       False          False  
8413           True       True      False       False          False  
2919           True      False      False       False          False  
6043          False      False      False       False          False  
...             ...        ...        ...         ...            ...  
13123          True       True      False       False          False  
3264          False      False      False       False          False  
9845          False      False      False       False          False  
10799          True       True      False       False          False  
2732          False      False      False       False          False  

[10864 rows x 12 columns]

### STEP 2


`.any()` gives if any one of records is `True`, we can pass an argument `axis=0` to perform this operation on each columns. If we used `axis=1` instead - we would have gotten missing values for each row. But we are looking for colums with missing values - so we use `axis=0` instead. 

I would encourage you to try both methods , axis = 0 and axis = 1 to see what happens. 

In [5]:
X_train.isnull().any(axis=0)

Rooms            False
Distance         False
Postcode         False
Bedroom2         False
Bathroom         False
Car               True
Landsize         False
BuildingArea      True
YearBuilt         True
Lattitude        False
Longtitude       False
Propertycount    False
dtype: bool

### STEP 3

Drop columns with missing values in training and validation data 

In [6]:
cols_with_missing = ['Car','BuildingArea','YearBuilt']

reduced_X_train = X_train.drop(cols_with_missing, axis=1)
reduced_X_valid = X_valid.drop(cols_with_missing, axis=1)

print("MAE from Approach 1 (Drop columns with missing values):")
print(score_dataset(reduced_X_train, reduced_X_valid, y_train, y_valid))

MAE from Approach 1 (Drop columns with missing values):
307246.80334151606


# Score from Approach 2 (Imputation)
***


<span style="color: #696969;">
    
Next, we use `SimpleImputer` to replace missing values with the `mean value` along each column.

Although it's simple, filling in the mean value generally performs quite well (but this varies by dataset). While statisticians have experimented with more complex ways to determine imputed values (such as regression imputation, for instance), the complex strategies typically give no additional benefit once you plug the results into sophisticated machine learning models.
    
<span>

In [7]:
from sklearn.impute import SimpleImputer

# Imputation
# Instantiate an instance of SimpleImputer
my_imputer = SimpleImputer()

# fit and transform each data set. 
# Note we are not providing any column names for imputation as the algorithm would apply this to all columns,
# and would transform only columns with missing vlaues.

imputed_X_train = pd.DataFrame(my_imputer.fit_transform(X_train))
imputed_X_valid = pd.DataFrame(my_imputer.transform(X_valid))

# Imputation removes column names; so we put them back, using column names from original datasets.
imputed_X_train.columns = X_train.columns
imputed_X_valid.columns = X_valid.columns

# use our function to calculate MAE from this approach
print("MAE from Approach 2 (Imputation):")
print(score_dataset(imputed_X_train, imputed_X_valid, y_train, y_valid))


MAE from Approach 2 (Imputation):
295580.75673035975


***
<span style="color: #696969;">
We see that `Approach 2` has lower MAE than `Approach 1`, so Approach 2 `performed better` on this dataset.

    
# Score from Approach 3 (An Extension to Imputation ~ OPTIONAL as I dont use it often)

Next, we impute the missing values, while also keeping track of which values were imputed.
</span>

In [8]:
# Make copy to avoid changing original data (when imputing)
X_train_plus = X_train.copy()
X_valid_plus = X_valid.copy()

In [9]:
# remember we created cols_with_missing variable giving list of all columns with missing values, lets use this again
cols_with_missing

['Car', 'BuildingArea', 'YearBuilt']

In [10]:
# Make new columns indicating what will be imputed,
# I have created a new column for each column with missing values for ex for car, i created a new columns 'car_was_missing'
# In this new column I filled True for wherever there was a missin value by using .isnull()

X_train_plus['Car_was_missing'] = X_train_plus['Car'].isnull()
X_valid_plus['Car_was_missing'] = X_valid_plus['Car'].isnull()

X_train_plus['BuildingArea_was_missing'] = X_train_plus['BuildingArea'].isnull()
X_valid_plus['BuildingArea_was_missing'] = X_valid_plus['BuildingArea'].isnull()

X_train_plus['YearBuilt_was_missing'] = X_train_plus['YearBuilt'].isnull()
X_valid_plus['YearBuilt_was_missing'] = X_valid_plus['YearBuilt'].isnull()


# You could have done all of above using a simple loop, if you are comfortable with loops have a look here 
######################################################################

# for col in cols_with_missing:
#    X_train_plus[col + '_was_missing'] = X_train_plus[col].isnull()
#    X_valid_plus[col + '_was_missing'] = X_valid_plus[col].isnull()

#####################################################################

In [11]:
X_train_plus

Rooms  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
12167      1       5.0    3182.0       1.0       1.0  1.0       0.0   
6524       2       8.0    3016.0       2.0       2.0  1.0     193.0   
8413       3      12.6    3020.0       3.0       1.0  1.0     555.0   
2919       3      13.0    3046.0       3.0       1.0  1.0     265.0   
6043       3      13.3    3020.0       3.0       1.0  2.0     673.0   
...      ...       ...       ...       ...       ...  ...       ...   
13123      3       5.2    3056.0       3.0       1.0  2.0     212.0   
3264       3      10.5    3081.0       3.0       1.0  1.0     748.0   
9845       4       6.7    3058.0       4.0       2.0  2.0     441.0   
10799      3      12.0    3073.0       3.0       1.0  1.0     606.0   
2732       4       6.4    3011.0       4.0       2.0  1.0     319.0   

       BuildingArea  YearBuilt  Lattitude  Longtitude  Propertycount  \
12167           NaN     1940.0  -37.85984   144.98670        13240.0   
6524            NaN        NaN  -37.85800   144.90050         6380.0   
8413            NaN        NaN  -37.79880   144.82200         3755.0   
2919            NaN     1995.0  -37.70830   144.91580         8870.0   
6043          673.0     1970.0  -37.76230   144.82720         4217.0   
...             ...        ...        ...         ...            ...   
13123           NaN        NaN  -37.77695   144.95785        11918.0   
3264          101.0     1950.0  -37.74160   145.04810         2947.0   
9845          255.0     2002.0  -37.73572   144.97256        11204.0   
10799           NaN        NaN  -37.72057   145.02615        21650.0   
2732          130.0     1915.0  -37.79430   144.88750         7570.0   

       Car_was_missing  BuildingArea_was_missing  YearBuilt_was_missing  
12167            False                      True                  False  
6524             False                      True                   True  
8413             False                      True                   True  
2919             False                      True                  False  
6043             False                     False                  False  
...                ...                       ...                    ...  
13123            False                      True                   True  
3264             False                     False                  False  
9845             False                     False                  False  
10799            False                      True                   True  
2732             False                     False                  False  

[10864 rows x 15 columns]

In [12]:
# Imputation
my_imputer = SimpleImputer()
imputed_X_train_plus = pd.DataFrame(my_imputer.fit_transform(X_train_plus))
imputed_X_valid_plus = pd.DataFrame(my_imputer.transform(X_valid_plus))

# Imputation removed column names; put them back
imputed_X_train_plus.columns = X_train_plus.columns
imputed_X_valid_plus.columns = X_valid_plus.columns

print("MAE from Approach 3 (An Extension to Imputation):")
print(score_dataset(imputed_X_train_plus, imputed_X_valid_plus, y_train, y_valid))

MAE from Approach 3 (An Extension to Imputation):
295735.91721870017


***
<span style="color: #696969;">

As we can see, Approach 3 performed allmost equal Approach 2.


`So, why did imputation perform better than dropping the columns?`

The training data has 10864 rows and 12 columns, where three columns contain missing data. For each column, less than half of the entries are missing. Thus, dropping the columns removes a lot of useful information, and so it makes sense that imputation would perform better.
</span>

***

In [13]:
# Shape of training data (num_rows, num_columns)

print(X_train.shape)

# Number of missing values in each column of training data
missing_val_count_by_column = (X_train.isnull().sum())
print(missing_val_count_by_column)

(10864, 12)
Rooms               0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                49
Landsize            0
BuildingArea     5156
YearBuilt        4307
Lattitude           0
Longtitude          0
Propertycount       0
dtype: int64


In [14]:
# we can write better code , lets just see the colums with missing values by subsetting the dataframe
missing_val_count_by_column[missing_val_count_by_column>0]

Car               49
BuildingArea    5156
YearBuilt       4307
dtype: int64

# APPROACH 4 : Most useful
<span style="color: #696969;">
This method we will try to break our entire dataset into blocks or stratas and impute mean for each block and use this mean to fill missing values in the block. For this we follow these steps :
</span>


### STEP 1: Add back the predicted value - since we would break our dataset into chunks using this predicted value

In [15]:
X_train['price'] = y_train

<ipython-input-15-ee5a4b415630>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_train['price'] = y_train


In [16]:
X_train

Rooms  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
12167      1       5.0    3182.0       1.0       1.0  1.0       0.0   
6524       2       8.0    3016.0       2.0       2.0  1.0     193.0   
8413       3      12.6    3020.0       3.0       1.0  1.0     555.0   
2919       3      13.0    3046.0       3.0       1.0  1.0     265.0   
6043       3      13.3    3020.0       3.0       1.0  2.0     673.0   
...      ...       ...       ...       ...       ...  ...       ...   
13123      3       5.2    3056.0       3.0       1.0  2.0     212.0   
3264       3      10.5    3081.0       3.0       1.0  1.0     748.0   
9845       4       6.7    3058.0       4.0       2.0  2.0     441.0   
10799      3      12.0    3073.0       3.0       1.0  1.0     606.0   
2732       4       6.4    3011.0       4.0       2.0  1.0     319.0   

       BuildingArea  YearBuilt  Lattitude  Longtitude  Propertycount  \
12167           NaN     1940.0  -37.85984   144.98670        13240.0   
6524            NaN        NaN  -37.85800   144.90050         6380.0   
8413            NaN        NaN  -37.79880   144.82200         3755.0   
2919            NaN     1995.0  -37.70830   144.91580         8870.0   
6043          673.0     1970.0  -37.76230   144.82720         4217.0   
...             ...        ...        ...         ...            ...   
13123           NaN        NaN  -37.77695   144.95785        11918.0   
3264          101.0     1950.0  -37.74160   145.04810         2947.0   
9845          255.0     2002.0  -37.73572   144.97256        11204.0   
10799           NaN        NaN  -37.72057   145.02615        21650.0   
2732          130.0     1915.0  -37.79430   144.88750         7570.0   

           price  
12167   481000.0  
6524    895000.0  
8413    651500.0  
2919    482500.0  
6043    591000.0  
...          ...  
13123  1280000.0  
3264    915000.0  
9845   1020000.0  
10799   760000.0  
2732   1225000.0  

[10864 rows x 13 columns]

***
## STEP 2: our Predicted value is contineous in nature so we break this into BINS. Lets say we break it to 10 BINS

<span style="color: #696969;">
I would recomend read the doc for pd.cut by using `alt+tab` after opening the bracket after ~ `pd.cut(alt+tab`
  
    
I have first used pd.cut to cut X_train['price'] into 5 buckets (ideally given the range of values, we should have cut this into 10 buckets but for sake of brevity, i am cutting this into 5 parts only) and labled each bucket with a number, try removing the lables argument and see what happens to X_train. Dont be afraid to experiment
</span>

In [17]:
X_train['price_bins'] = pd.cut(X_train['price'],5,labels=[1,2,3,4,5])

<ipython-input-17-b7e5def53746>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_train['price_bins'] = pd.cut(X_train['price'],5,labels=[1,2,3,4,5])


In [18]:
X_train

Rooms  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
12167      1       5.0    3182.0       1.0       1.0  1.0       0.0   
6524       2       8.0    3016.0       2.0       2.0  1.0     193.0   
8413       3      12.6    3020.0       3.0       1.0  1.0     555.0   
2919       3      13.0    3046.0       3.0       1.0  1.0     265.0   
6043       3      13.3    3020.0       3.0       1.0  2.0     673.0   
...      ...       ...       ...       ...       ...  ...       ...   
13123      3       5.2    3056.0       3.0       1.0  2.0     212.0   
3264       3      10.5    3081.0       3.0       1.0  1.0     748.0   
9845       4       6.7    3058.0       4.0       2.0  2.0     441.0   
10799      3      12.0    3073.0       3.0       1.0  1.0     606.0   
2732       4       6.4    3011.0       4.0       2.0  1.0     319.0   

       BuildingArea  YearBuilt  Lattitude  Longtitude  Propertycount  \
12167           NaN     1940.0  -37.85984   144.98670        13240.0   
6524            NaN        NaN  -37.85800   144.90050         6380.0   
8413            NaN        NaN  -37.79880   144.82200         3755.0   
2919            NaN     1995.0  -37.70830   144.91580         8870.0   
6043          673.0     1970.0  -37.76230   144.82720         4217.0   
...             ...        ...        ...         ...            ...   
13123           NaN        NaN  -37.77695   144.95785        11918.0   
3264          101.0     1950.0  -37.74160   145.04810         2947.0   
9845          255.0     2002.0  -37.73572   144.97256        11204.0   
10799           NaN        NaN  -37.72057   145.02615        21650.0   
2732          130.0     1915.0  -37.79430   144.88750         7570.0   

           price price_bins  
12167   481000.0          1  
6524    895000.0          1  
8413    651500.0          1  
2919    482500.0          1  
6043    591000.0          1  
...          ...        ...  
13123  1280000.0          1  
3264    915000.0          1  
9845   1020000.0          1  
10799   760000.0          1  
2732   1225000.0          1  

[10864 rows x 14 columns]

In [19]:
# Columns with missing values
cols_with_missing

['Car', 'BuildingArea', 'YearBuilt']

***
### STEP 3: 

<span style="color: #696969;">
Get Mean values for each bucket. We do this one by one for each bucket and for each column with missing data
</span>

In [20]:
Group1_car = X_train[X_train['price_bins']==1]['Car'].mean()
Group2_car = X_train[X_train['price_bins']==2]['Car'].mean()
Group3_car = X_train[X_train['price_bins']==3]['Car'].mean()
Group4_car = X_train[X_train['price_bins']==4]['Car'].mean()
Group5_car = X_train[X_train['price_bins']==5]['Car'].mean()


print('mean of car for group1 is: {}'.format(Group1_car))
print('mean of car for group2 is: {}'.format(Group2_car))
print('mean of car for group3 is: {}'.format(Group3_car))
print('mean of car for group4 is: {}'.format(Group4_car))
print('mean of car for group5 is: {}\n'.format(Group5_car))

Group1_BuildingArea = X_train[X_train['price_bins']==1]['BuildingArea'].mean()
Group2_BuildingArea = X_train[X_train['price_bins']==2]['BuildingArea'].mean()
Group3_BuildingArea = X_train[X_train['price_bins']==3]['BuildingArea'].mean()
Group4_BuildingArea = X_train[X_train['price_bins']==4]['BuildingArea'].mean()
Group5_BuildingArea = X_train[X_train['price_bins']==5]['BuildingArea'].mean()

print('mean of BuildingArea for group1 is: {}'.format(Group1_BuildingArea))
print('mean of BuildingArea for group2 is: {}'.format(Group2_BuildingArea))
print('mean of BuildingArea for group3 is: {}'.format(Group3_BuildingArea))
print('mean of BuildingArea for group4 is: {}'.format(Group4_BuildingArea))
print('mean of BuildingArea for group5 is: {}\n'.format(Group5_BuildingArea))

Group1_YearBuilt = X_train[X_train['price_bins']==1]['YearBuilt'].mean()
Group2_YearBuilt = X_train[X_train['price_bins']==2]['YearBuilt'].mean()
Group3_YearBuilt = X_train[X_train['price_bins']==3]['YearBuilt'].mean()
Group4_YearBuilt = X_train[X_train['price_bins']==4]['YearBuilt'].mean()
Group5_YearBuilt = X_train[X_train['price_bins']==5]['YearBuilt'].mean()

print('mean of YearBuilt for group1 is: {}'.format(Group1_YearBuilt))
print('mean of YearBuilt for group2 is: {}'.format(Group2_YearBuilt))
print('mean of YearBuilt for group3 is: {}'.format(Group3_YearBuilt))
print('mean of YearBuilt for group4 is: {}'.format(Group4_YearBuilt))
print('mean of YearBuilt for group5 is: {}\n'.format(Group5_YearBuilt))

mean of car for group1 is: 1.5565671031096564
mean of car for group2 is: 2.017708333333333
mean of car for group3 is: 2.5
mean of car for group4 is: 2.75
mean of car for group5 is: 3.0

mean of BuildingArea for group1 is: 142.4214511971969
mean of BuildingArea for group2 is: 250.77389847619048
mean of BuildingArea for group3 is: 306.5945945945946
mean of BuildingArea for group4 is: 366.8333333333333
mean of BuildingArea for group5 is: 288.43333333333334

mean of YearBuilt for group1 is: 1967.2619289340103
mean of YearBuilt for group2 is: 1942.8825503355704
mean of YearBuilt for group3 is: 1945.560975609756
mean of YearBuilt for group4 is: 1929.7142857142858
mean of YearBuilt for group5 is: 1901.0



***
### STEP 4:

<span style="color: #696969;">
replace missing value for each strata by mean of that strata.    
</span>

Break down of the formulae i have used :

1. `X_train.loc` : this takes location of index to subste the data (read the documentation once ). this takes two arguments first one is row and second column, I have shown both of these below in 2 & 3
2. `(X_train['Car'].isnull()) & (X_train['price_bins']==1)` : this part gives all rows with car as missing `AND` 
pricebin = 1
3. `"Car"` : This part subsets only the column car from the dataframe 
4. This sub-setted dataframe is given value of group1_car which we calculated above

In [21]:
X_train.loc[(X_train['Car'].isnull()) & (X_train['price_bins']==1),"Car"] = Group1_car
X_train.loc[(X_train['Car'].isnull()) & (X_train['price_bins']==2),"Car"] = Group2_car
X_train.loc[(X_train['Car'].isnull()) & (X_train['price_bins']==3),"Car"] = Group3_car
X_train.loc[(X_train['Car'].isnull()) & (X_train['price_bins']==4),"Car"] = Group4_car
X_train.loc[(X_train['Car'].isnull()) & (X_train['price_bins']==5),"Car"] = Group5_car

X_train.loc[(X_train['BuildingArea'].isnull()) & (X_train['price_bins']==1),"BuildingArea"] = Group1_BuildingArea
X_train.loc[(X_train['BuildingArea'].isnull()) & (X_train['price_bins']==2),"BuildingArea"] = Group2_BuildingArea
X_train.loc[(X_train['BuildingArea'].isnull()) & (X_train['price_bins']==3),"BuildingArea"] = Group3_BuildingArea
X_train.loc[(X_train['BuildingArea'].isnull()) & (X_train['price_bins']==4),"BuildingArea"] = Group4_BuildingArea
X_train.loc[(X_train['BuildingArea'].isnull()) & (X_train['price_bins']==5),"BuildingArea"] = Group5_BuildingArea

X_train.loc[(X_train['YearBuilt'].isnull()) & (X_train['price_bins']==1),"YearBuilt"] = Group1_YearBuilt
X_train.loc[(X_train['YearBuilt'].isnull()) & (X_train['price_bins']==2),"YearBuilt"] = Group2_YearBuilt
X_train.loc[(X_train['YearBuilt'].isnull()) & (X_train['price_bins']==3),"YearBuilt"] = Group3_YearBuilt
X_train.loc[(X_train['YearBuilt'].isnull()) & (X_train['price_bins']==4),"YearBuilt"] = Group4_YearBuilt
X_train.loc[(X_train['YearBuilt'].isnull()) & (X_train['price_bins']==5),"YearBuilt"] = Group5_YearBuilt

/home/bharath/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s


In [22]:
# check if ther are any null values
X_train.isnull().sum()

Rooms            0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
YearBuilt        0
Lattitude        0
Longtitude       0
Propertycount    0
price            0
price_bins       0
dtype: int64

In [23]:
X_train

Rooms  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
12167      1       5.0    3182.0       1.0       1.0  1.0       0.0   
6524       2       8.0    3016.0       2.0       2.0  1.0     193.0   
8413       3      12.6    3020.0       3.0       1.0  1.0     555.0   
2919       3      13.0    3046.0       3.0       1.0  1.0     265.0   
6043       3      13.3    3020.0       3.0       1.0  2.0     673.0   
...      ...       ...       ...       ...       ...  ...       ...   
13123      3       5.2    3056.0       3.0       1.0  2.0     212.0   
3264       3      10.5    3081.0       3.0       1.0  1.0     748.0   
9845       4       6.7    3058.0       4.0       2.0  2.0     441.0   
10799      3      12.0    3073.0       3.0       1.0  1.0     606.0   
2732       4       6.4    3011.0       4.0       2.0  1.0     319.0   

       BuildingArea    YearBuilt  Lattitude  Longtitude  Propertycount  \
12167    142.421451  1940.000000  -37.85984   144.98670        13240.0   
6524     142.421451  1967.261929  -37.85800   144.90050         6380.0   
8413     142.421451  1967.261929  -37.79880   144.82200         3755.0   
2919     142.421451  1995.000000  -37.70830   144.91580         8870.0   
6043     673.000000  1970.000000  -37.76230   144.82720         4217.0   
...             ...          ...        ...         ...            ...   
13123    142.421451  1967.261929  -37.77695   144.95785        11918.0   
3264     101.000000  1950.000000  -37.74160   145.04810         2947.0   
9845     255.000000  2002.000000  -37.73572   144.97256        11204.0   
10799    142.421451  1967.261929  -37.72057   145.02615        21650.0   
2732     130.000000  1915.000000  -37.79430   144.88750         7570.0   

           price price_bins  
12167   481000.0          1  
6524    895000.0          1  
8413    651500.0          1  
2919    482500.0          1  
6043    591000.0          1  
...          ...        ...  
13123  1280000.0          1  
3264    915000.0          1  
9845   1020000.0          1  
10799   760000.0          1  
2732   1225000.0          1  

[10864 rows x 14 columns]

***
### STEP 5:

<span style="color: #696969;">
Remove `price` and `price_bins` from the dataframe as we no long need them
</span>

In [24]:
X_train.drop(['price','price_bins'],axis=1,inplace=True)

/home/bharath/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:3990: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [25]:
X_train

Rooms  Distance  Postcode  Bedroom2  Bathroom  Car  Landsize  \
12167      1       5.0    3182.0       1.0       1.0  1.0       0.0   
6524       2       8.0    3016.0       2.0       2.0  1.0     193.0   
8413       3      12.6    3020.0       3.0       1.0  1.0     555.0   
2919       3      13.0    3046.0       3.0       1.0  1.0     265.0   
6043       3      13.3    3020.0       3.0       1.0  2.0     673.0   
...      ...       ...       ...       ...       ...  ...       ...   
13123      3       5.2    3056.0       3.0       1.0  2.0     212.0   
3264       3      10.5    3081.0       3.0       1.0  1.0     748.0   
9845       4       6.7    3058.0       4.0       2.0  2.0     441.0   
10799      3      12.0    3073.0       3.0       1.0  1.0     606.0   
2732       4       6.4    3011.0       4.0       2.0  1.0     319.0   

       BuildingArea    YearBuilt  Lattitude  Longtitude  Propertycount  
12167    142.421451  1940.000000  -37.85984   144.98670        13240.0  
6524     142.421451  1967.261929  -37.85800   144.90050         6380.0  
8413     142.421451  1967.261929  -37.79880   144.82200         3755.0  
2919     142.421451  1995.000000  -37.70830   144.91580         8870.0  
6043     673.000000  1970.000000  -37.76230   144.82720         4217.0  
...             ...          ...        ...         ...            ...  
13123    142.421451  1967.261929  -37.77695   144.95785        11918.0  
3264     101.000000  1950.000000  -37.74160   145.04810         2947.0  
9845     255.000000  2002.000000  -37.73572   144.97256        11204.0  
10799    142.421451  1967.261929  -37.72057   145.02615        21650.0  
2732     130.000000  1915.000000  -37.79430   144.88750         7570.0  

[10864 rows x 12 columns]

### STEP 6:

<span style="color: #696969;">
Repeat all above steps for validation set

</span>

In [26]:
# Step 1 : add back price to X valid dataset
X_valid['price'] = y_valid

# Step 2: break X validation set to 5 bins
X_valid['price_bins'] = pd.cut(X_valid['price'],5,labels=[1,2,3,4,5])

# Step 3: calculate mean of each bin
Group1_car = X_valid[X_valid['price_bins']==1]['Car'].mean()
Group2_car = X_valid[X_valid['price_bins']==2]['Car'].mean()
Group3_car = X_valid[X_valid['price_bins']==3]['Car'].mean()
Group4_car = X_valid[X_valid['price_bins']==4]['Car'].mean()
Group5_car = X_valid[X_valid['price_bins']==5]['Car'].mean()

Group1_BuildingArea = X_valid[X_valid['price_bins']==1]['BuildingArea'].mean()
Group2_BuildingArea = X_valid[X_valid['price_bins']==2]['BuildingArea'].mean()
Group3_BuildingArea = X_valid[X_valid['price_bins']==3]['BuildingArea'].mean()
Group4_BuildingArea = X_valid[X_valid['price_bins']==4]['BuildingArea'].mean()
Group5_BuildingArea = X_valid[X_valid['price_bins']==5]['BuildingArea'].mean()

Group1_YearBuilt = X_valid[X_valid['price_bins']==1]['YearBuilt'].mean()
Group2_YearBuilt = X_valid[X_valid['price_bins']==2]['YearBuilt'].mean()
Group3_YearBuilt = X_valid[X_valid['price_bins']==3]['YearBuilt'].mean()
Group4_YearBuilt = X_valid[X_valid['price_bins']==4]['YearBuilt'].mean()
Group5_YearBuilt = X_valid[X_valid['price_bins']==5]['YearBuilt'].mean()

# Step 4: replace missing values for each strata
X_valid.loc[(X_valid['Car'].isnull()) & (X_valid['price_bins']==1),"Car"] = Group1_car
X_valid.loc[(X_valid['Car'].isnull()) & (X_valid['price_bins']==2),"Car"] = Group2_car
X_valid.loc[(X_valid['Car'].isnull()) & (X_valid['price_bins']==3),"Car"] = Group3_car
X_valid.loc[(X_valid['Car'].isnull()) & (X_valid['price_bins']==4),"Car"] = Group4_car
X_valid.loc[(X_valid['Car'].isnull()) & (X_valid['price_bins']==5),"Car"] = Group5_car

X_valid.loc[(X_valid['BuildingArea'].isnull()) & (X_valid['price_bins']==1),"BuildingArea"] = Group1_BuildingArea
X_valid.loc[(X_valid['BuildingArea'].isnull()) & (X_valid['price_bins']==2),"BuildingArea"] = Group2_BuildingArea
X_valid.loc[(X_valid['BuildingArea'].isnull()) & (X_valid['price_bins']==3),"BuildingArea"] = Group3_BuildingArea
X_valid.loc[(X_valid['BuildingArea'].isnull()) & (X_valid['price_bins']==4),"BuildingArea"] = Group4_BuildingArea
X_valid.loc[(X_valid['BuildingArea'].isnull()) & (X_valid['price_bins']==5),"BuildingArea"] = Group5_BuildingArea

X_valid.loc[(X_valid['YearBuilt'].isnull()) & (X_valid['price_bins']==1),"YearBuilt"] = Group1_YearBuilt
X_valid.loc[(X_valid['YearBuilt'].isnull()) & (X_valid['price_bins']==2),"YearBuilt"] = Group2_YearBuilt
X_valid.loc[(X_valid['YearBuilt'].isnull()) & (X_valid['price_bins']==3),"YearBuilt"] = Group3_YearBuilt
X_valid.loc[(X_valid['YearBuilt'].isnull()) & (X_valid['price_bins']==4),"YearBuilt"] = Group4_YearBuilt
X_valid.loc[(X_valid['YearBuilt'].isnull()) & (X_valid['price_bins']==5),"YearBuilt"] = Group5_YearBuilt

# Step 5: remove Price and Price bin columns
X_valid.drop(['price','price_bins'],axis=1,inplace=True)

<ipython-input-26-5c762400c61d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_valid['price'] = y_valid
<ipython-input-26-5c762400c61d>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_valid['price_bins'] = pd.cut(X_valid['price'],5,labels=[1,2,3,4,5])
/home/bharath/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:966: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the 

### STEP 7:

<span style="color: #696969;">
Calculate the MAE for the dataset with stratified imputation

</span>

In [27]:
print("MAE from Approach 3 (An Extension to Imputation):")
print(score_dataset(X_train, X_valid, y_train, y_valid))

MAE from Approach 3 (An Extension to Imputation):
287699.5074855098


# This is the best score by far, one should go for stratified imputation where ever the datasets have disproportional representation is each strata

# KNN IMPUTER

In [28]:
from sklearn.impute import KNNImputer

In [29]:
myimputer = KNNImputer(n_neighbors=10)

In [30]:
# impute using KNN algorithm 
imputed_X_train = myimputer.fit_transform(X_train)
# add back columns names as imputer removes column names.
#imputed_X_train.columns = X_train.columns

# do same for validation set.
imputed_X_valid = myimputer.fit_transform(X_valid)
#imputed_X_valid.columns = X_valid.columns

In [31]:
print("MAE from Approach 3 (An Extension to Imputation):")
print(score_dataset(imputed_X_train, imputed_X_valid, y_train, y_valid))

MAE from Approach 3 (An Extension to Imputation):
287699.5074855098


***
we should use KNN imputer when datasets are not imbalanced but there is a lot of variance in data. 

***

***

# Conclusion

<span style="color: #696969;">
As is common, imputing missing values (in Approach 2 and Approach 3) yielded better results, relative to when we simply dropped columns with missing values (in Approach 1).
    
Stratified Imputation should allways be used if dataset is stratified. You can check for stratification during your uni-variate analyis step of EDA
    
For non stratified data but data with large variance - prefer KNN imputation.
</span>

# Your Turn

<span style="color: #696969;">
Compare these approaches to dealing with missing values yourself in this exercise!
</span>

# End of notebook